pyseir/models/historical_data.py

```diff
@@ -0,0 +1,220 @@
+        rows["nC"] = rows["positive_tests"] - rows["positive_tests"].shift(fill_value=0.0)
+
+        cum_tests = rows["positive_tests"] + rows["negative_tests"]
+        rows["tests"] = cum_tests - cum_tests.shift(fill_value=0.0)
+
+        rows["H"] = rows["current_hospitalized"]
+        rows["nD"] = rows["deaths"] - rows["deaths"].shift(fill_value=0.0)
+
+        # Apply fixes before smoothing
+        if state in DATA_CORRECTIONS:
+            state_corrections = DATA_CORRECTIONS[state]
+            for col in state_corrections:
+                for (day, value) in state_corrections[col]:
+                    rows[col][day] = value
+
+        # Then smooth the data
+        rows["nC"] = HistoricalData.smooth(rows["nC"])
+        rows["tests"] = HistoricalData.smooth(rows["tests"])
+        rows["H"] = HistoricalData.smooth(rows["H"])
+        rows["nD"] = HistoricalData.smooth(rows["nD"])
+
+        # Apply date filter last
+        early_rows = rows[(rows["date"] >= early_start) & (rows["date"] <= end_date)]
+        rows = early_rows[(early_rows["date"] >= start_date)]
+
+        if compartments_as_functions:
+            return (
+                lambda t: early_rows["rt"][t],
+                lambda t: early_rows["nC"][t],
+                lambda t: early_rows["tests"][t],
+                lambda t: early_rows["H"][t],
+                lambda t: early_rows["nD"][t],
+            )
+        elif no_functions:
+            return (
+                rows["rt"],
+                rows["nC"],
+                rows["tests"],
+                rows["H"],
+                rows["nD"],
+            )
+        else:
+            return (
+                lambda t: early_rows["rt"][t],
+                rows["nC"],
+                lambda t: early_rows["tests"][t],
+                rows["H"],
+                rows["nD"],
+            )
+
+    @staticmethod
+    def get_states():
+        return HistoricalData.data["state"].unique()
+
+
+def adjust_rt_to_match_cases(rt_f, new_cases, t_list):
+    """
+    Given R(t) and new_cases(t) series and a time range, find adjustment to R(t)
+    so that it properly predicts the ratio of ending/starting new_cases
+    """
+    rt = [rt_f(t) for t in t_list]
+    average_R = sum(rt) / len(rt)
+    case_growth = new_cases(t_list[-1]) / new_cases(t_list[0])
+    r_1_growth = math.exp((average_R - 1) * (len(t_list) - 1) / InferRtConstants.SERIAL_PERIOD)
+    growth_ratio = case_growth / r_1_growth
+    if growth_ratio > 0:
+        dr = math.log(growth_ratio) * InferRtConstants.SERIAL_PERIOD / (len(t_list) - 1)
+    else:  # something funny so skip
+        dr = 0.0
+    adj_r_f = lambda t: rt_f(t) + dr
+
+    return (average_R, growth_ratio, adj_r_f)
+
+
+class ForecastData:
+    """
+    Look up forecast data to use for testing purposes.
+    """
+
+    raw = pd.read_csv(FORECAST_DATA_FILE, parse_dates=["date"])
+    data = raw[["fips", "date", "weekly_new_cases_0.5", "weekly_new_deaths_0.5",]]
```

Comment:
what is the weekly_new_cases_0.5 mean?

Reply:
Added comments explaining the .5 means median value from a set of projections